{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Module_2_Test.ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyMnJwD2f0O5DgIqyZSabAuq"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["### Module 2 Test"],"metadata":{"id":"BWFmZwjD7Qqp"}},{"cell_type":"markdown","source":["### Read Genome Function"],"metadata":{"id":"WQPOvcY6HZCy"}},{"cell_type":"code","source":["def readGenome(filename):\n","    genome = ''\n","    with open(filename, 'r') as f:\n","        for line in f:\n","            # ignore header line with genome information\n","            if not line[0] == '>':\n","                genome += line.rstrip()\n","    return genome"],"metadata":{"id":"9_t0W4AtHds2","executionInfo":{"status":"ok","timestamp":1656618630741,"user_tz":-60,"elapsed":170,"user":{"displayName":"Sasha malkani","userId":"10596222964619627791"}}},"execution_count":50,"outputs":[]},{"cell_type":"markdown","source":["### Python Module for Boyer-Moore Pre-processing"],"metadata":{"id":"pJbOyBg17Z09"}},{"cell_type":"code","execution_count":51,"metadata":{"id":"ioT99i8a7Lkz","executionInfo":{"status":"ok","timestamp":1656618631549,"user_tz":-60,"elapsed":511,"user":{"displayName":"Sasha malkani","userId":"10596222964619627791"}}},"outputs":[],"source":["\"\"\"bm_preproc.py: Boyer-Moore preprocessing.\"\"\"\n","\n","\"\"\"__author__ = \"Ben Langmead\" \"\"\"\n","\n","import unittest\n","\n","\n","def z_array(s):\n","    \"\"\" Use Z algorithm (Gusfield theorem 1.4.1) to preprocess s \"\"\"\n","    assert len(s) > 1\n","    z = [len(s)] + [0] * (len(s)-1)\n","\n","    # Initial comparison of s[1:] with prefix\n","    for i in range(1, len(s)):\n","        if s[i] == s[i-1]:\n","            z[1] += 1\n","        else:\n","            break\n","\n","    r, l = 0, 0\n","    if z[1] > 0:\n","        r, l = z[1], 1\n","\n","    for k in range(2, len(s)):\n","        assert z[k] == 0\n","        if k > r:\n","            # Case 1\n","            for i in range(k, len(s)):\n","                if s[i] == s[i-k]:\n","                    z[k] += 1\n","                else:\n","                    break\n","            r, l = k + z[k] - 1, k\n","        else:\n","            # Case 2\n","            # Calculate length of beta\n","            nbeta = r - k + 1\n","            zkp = z[k - l]\n","            if nbeta > zkp:\n","                # Case 2a: zkp wins\n","                z[k] = zkp\n","            else:\n","                # Case 2b: Compare characters just past r\n","                nmatch = 0\n","                for i in range(r+1, len(s)):\n","                    if s[i] == s[i - k]:\n","                        nmatch += 1\n","                    else:\n","                        break\n","                l, r = k, r + nmatch\n","                z[k] = r - k + 1\n","    return z\n","\n","\n","def n_array(s):\n","    \"\"\" Compile the N array (Gusfield theorem 2.2.2) from the Z array \"\"\"\n","    return z_array(s[::-1])[::-1]\n","\n","\n","def big_l_prime_array(p, n):\n","    \"\"\" Compile L' array (Gusfield theorem 2.2.2) using p and N array.\n","        L'[i] = largest index j less than n such that N[j] = |P[i:]| \"\"\"\n","    lp = [0] * len(p)\n","    for j in range(len(p)-1):\n","        i = len(p) - n[j]\n","        if i < len(p):\n","            lp[i] = j + 1\n","    return lp\n","\n","\n","def big_l_array(p, lp):\n","    \"\"\" Compile L array (Gusfield theorem 2.2.2) using p and L' array.\n","        L[i] = largest index j less than n such that N[j] >= |P[i:]| \"\"\"\n","    l = [0] * len(p)\n","    l[1] = lp[1]\n","    for i in range(2, len(p)):\n","        l[i] = max(l[i-1], lp[i])\n","    return l\n","\n","\n","def small_l_prime_array(n):\n","    \"\"\" Compile lp' array (Gusfield theorem 2.2.4) using N array. \"\"\"\n","    small_lp = [0] * len(n)\n","    for i in range(len(n)):\n","        if n[i] == i+1:  # prefix matching a suffix\n","            small_lp[len(n)-i-1] = i+1\n","    for i in range(len(n)-2, -1, -1):  # \"smear\" them out to the left\n","        if small_lp[i] == 0:\n","            small_lp[i] = small_lp[i+1]\n","    return small_lp\n","\n","\n","def good_suffix_table(p):\n","    \"\"\" Return tables needed to apply good suffix rule. \"\"\"\n","    n = n_array(p)\n","    lp = big_l_prime_array(p, n)\n","    return lp, big_l_array(p, lp), small_l_prime_array(n)\n","\n","\n","def good_suffix_mismatch(i, big_l_prime, small_l_prime):\n","    \"\"\" Given a mismatch at offset i, and given L/L' and l' arrays,\n","        return amount to shift as determined by good suffix rule. \"\"\"\n","    length = len(big_l_prime)\n","    assert i < length\n","    if i == length - 1:\n","        return 0\n","    i += 1  # i points to leftmost matching position of P\n","    if big_l_prime[i] > 0:\n","        return length - big_l_prime[i]\n","    return length - small_l_prime[i]\n","\n","\n","def good_suffix_match(small_l_prime):\n","    \"\"\" Given a full match of P to T, return amount to shift as\n","        determined by good suffix rule. \"\"\"\n","    return len(small_l_prime) - small_l_prime[1]\n","\n","\n","def dense_bad_char_tab(p, amap):\n","    \"\"\" Given pattern string and list with ordered alphabet characters, create\n","        and return a dense bad character table.  Table is indexed by offset\n","        then by character. \"\"\"\n","    tab = []\n","    nxt = [0] * len(amap)\n","    for i in range(0, len(p)):\n","        c = p[i]\n","        assert c in amap\n","        tab.append(nxt[:])\n","        nxt[amap[c]] = i+1\n","    return tab\n","\n","\n","class BoyerMoore(object):\n","    \"\"\" Encapsulates pattern and associated Boyer-Moore preprocessing. \"\"\"\n","\n","    def __init__(self, p, alphabet='ACGT'):\n","        # Create map from alphabet characters to integers\n","        self.amap = {alphabet[i]: i for i in range(len(alphabet))}\n","        # Make bad character rule table\n","        self.bad_char = dense_bad_char_tab(p, self.amap)\n","        # Create good suffix rule table\n","        _, self.big_l, self.small_l_prime = good_suffix_table(p)\n","\n","    def bad_character_rule(self, i, c):\n","        \"\"\" Return # skips given by bad character rule at offset i \"\"\"\n","        assert c in self.amap\n","        assert i < len(self.bad_char)\n","        ci = self.amap[c]\n","        return i - (self.bad_char[i][ci]-1)\n","\n","    def good_suffix_rule(self, i):\n","        \"\"\" Given a mismatch at offset i, return amount to shift\n","            as determined by (weak) good suffix rule. \"\"\"\n","        length = len(self.big_l)\n","        assert i < length\n","        if i == length - 1:\n","            return 0\n","        i += 1  # i points to leftmost matching position of P\n","        if self.big_l[i] > 0:\n","            return length - self.big_l[i]\n","        return length - self.small_l_prime[i]\n","\n","    def match_skip(self):\n","        \"\"\" Return amount to shift in case where P matches T \"\"\"\n","        return len(self.small_l_prime) - self.small_l_prime[1]\n","\n","\n"]},{"cell_type":"markdown","source":["### Naive Exact Matching Algorithm"],"metadata":{"id":"5hK0NhAvBvzp"}},{"cell_type":"code","source":["def naive_with_counts(p, t):\n","    occurences = []\n","    count_chr = 0 # initialize counter for character comparisons\n","    count_align = 0 # initialize counter for alignments\n","    for i in range(len(t) - len(p) + 1):\n","        count_align += 1 # add 1 to count_align\n","        match = True\n","        for j in range(len(p)):\n","            count_chr += 1 # add 1 to count_chr\n","            if not t[i+j] == p[j]:\n","                match = False\n","                break\n","        if match:\n","            occurences.append(i)\n","    return occurences, count_align, count_chr"],"metadata":{"id":"Eh_mc2CrB2PG","executionInfo":{"status":"ok","timestamp":1656618631550,"user_tz":-60,"elapsed":7,"user":{"displayName":"Sasha malkani","userId":"10596222964619627791"}}},"execution_count":52,"outputs":[]},{"cell_type":"markdown","source":["### Boyer-Moore Function"],"metadata":{"id":"w6VHQa-FBBiT"}},{"cell_type":"code","source":["                                  # p=pattern, p_bm=Boyer-Moore Object, t=text\n","def boyer_moore_with_counts(p, p_bm, t): \n","    i = 0\n","    occurrences = []\n","    count_chr = 0                 # initialize counter for character comparisons\n","    count_align = 0               # initialize counter for alignments\n","    while i < len(t) - len(p) + 1:\n","        count_align += 1                  # add 1 to count_align\n","        # amount pattern p will move along text t for comparison can be >1\n","        shift = 1\n","        # will be updated if we find a mismatch                         \n","        mismatched = False \n","        for j in range(len(p)-1, -1, -1): # loop through text from back to front\n","            count_chr += 1                # add 1 to count_chr \n","            if not p[j] == t[i+j]:        # character comparison\n","                # calling bad character rule\n","                skip_bc = p_bm.bad_character_rule(j, t[i+j])\n","                # calling good suffix rule \n","                skip_gs = p_bm.good_suffix_rule(j)\n","                # move p along t by the max of either shift, skip_bc or skip_gs           \n","                shift = max(shift, skip_bc, skip_gs) \n","                mismatched = True         # mismatch found\n","                break                     # break loop if mismatch found\n","        if not mismatched:                # no mismatch, exact match of text t\n","            # add position onto list of occurences\n","            occurrences.append(i)\n","            # still skip some using match skip function         \n","            skip_gs = p_bm.match_skip()\n","            # move p along t by the max of either shift or skip_gs   \n","            shift = max(shift, skip_gs)   \n","        i += shift                        # update our position by shift\n","\n","    # Return occurrences, no. of alignments, no. of character comparisons\n","    return occurrences, count_align, count_chr "],"metadata":{"id":"oZvj_4dwBJoQ","executionInfo":{"status":"ok","timestamp":1656618631551,"user_tz":-60,"elapsed":7,"user":{"displayName":"Sasha malkani","userId":"10596222964619627791"}}},"execution_count":53,"outputs":[]},{"cell_type":"markdown","source":["### Downloading Human Chromosome 1"],"metadata":{"id":"LsNNYFF6I2Ho"}},{"cell_type":"code","source":["!wget --no-check http://d28rh4a8wq0iu5.cloudfront.net/ads1/data/chr1.GRCh38.excerpt.fasta"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"neMu6pW1I9Mr","executionInfo":{"status":"ok","timestamp":1656618632372,"user_tz":-60,"elapsed":828,"user":{"displayName":"Sasha malkani","userId":"10596222964619627791"}},"outputId":"b590dba4-c2de-494e-c31f-ea09d77ca7db"},"execution_count":54,"outputs":[{"output_type":"stream","name":"stdout","text":["--2022-06-30 19:50:30--  http://d28rh4a8wq0iu5.cloudfront.net/ads1/data/chr1.GRCh38.excerpt.fasta\n","Resolving d28rh4a8wq0iu5.cloudfront.net (d28rh4a8wq0iu5.cloudfront.net)... 108.156.105.76, 108.156.105.28, 108.156.105.8, ...\n","Connecting to d28rh4a8wq0iu5.cloudfront.net (d28rh4a8wq0iu5.cloudfront.net)|108.156.105.76|:80... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 810105 (791K) [application/octet-stream]\n","Saving to: ‘chr1.GRCh38.excerpt.fasta.1’\n","\n","\r          chr1.GRCh   0%[                    ]       0  --.-KB/s               \rchr1.GRCh38.excerpt 100%[===================>] 791.12K  --.-KB/s    in 0.05s   \n","\n","2022-06-30 19:50:31 (14.7 MB/s) - ‘chr1.GRCh38.excerpt.fasta.1’ saved [810105/810105]\n","\n"]}]},{"cell_type":"markdown","source":["### Read fasta file (downloaded Human Chromosome 1 (genome))"],"metadata":{"id":"acq7zmBcb5lW"}},{"cell_type":"code","source":["genome_file = 'chr1.GRCh38.excerpt.fasta'\n","\n","t = readGenome(genome_file)\n","\n","print(t[:100])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"jmnOX3PEcPJj","executionInfo":{"status":"ok","timestamp":1656618632373,"user_tz":-60,"elapsed":7,"user":{"displayName":"Sasha malkani","userId":"10596222964619627791"}},"outputId":"a8e42308-43ba-4f0b-f0da-7d5fb0308b96"},"execution_count":55,"outputs":[{"output_type":"stream","name":"stdout","text":["TTGAATGCTGAAATCAGCAGGTAATATATGATAATAGAGAAAGCTATCCCGAAGGTGCATAGGTCAACAATACTTGAGCCTAACTCAGTAGATCCTAAAA\n"]}]},{"cell_type":"markdown","source":["### Calculate the number of; occurences, alignments, character comparisons of pattern p in text (genome) t \n","### using the Naive Exact Algorithm"],"metadata":{"id":"ttrCEZaHdSSy"}},{"cell_type":"code","source":["p = 'GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG'\n","\n","occurrences, num_alignments, num_character_comparisons = naive_with_counts(p, t)\n","\n","print(\"Exact occurrences: %d\" % occurrences[0])\n","print(\"No. of alignments: %d\" % num_alignments)\n","print(\"No. of character comparisons: %d\" % num_character_comparisons)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"PZk1Sfu1eDzn","executionInfo":{"status":"ok","timestamp":1656618633179,"user_tz":-60,"elapsed":809,"user":{"displayName":"Sasha malkani","userId":"10596222964619627791"}},"outputId":"06d0c825-7087-44be-eb98-067c76666cd6"},"execution_count":56,"outputs":[{"output_type":"stream","name":"stdout","text":["Exact occurrences: 56922\n","No. of alignments: 799954\n","No. of character comparisons: 984143\n"]}]},{"cell_type":"markdown","source":["### Calculate the number of; occurences, alignments, character comparisons of pattern p in text (genome) t \n","### using the Boyer-Moore Algorithm"],"metadata":{"id":"3B5d4amLitDG"}},{"cell_type":"code","source":["p_bm = BoyerMoore(p, alphabet='ACGT')\n","\n","occurrences, num_align, num_char_compar = boyer_moore_with_counts(p, p_bm, t)\n","\n","print(\"Exact occurrences: %d\" % occurrences[0])\n","print(\"No. of alignments: %d\" % num_align)\n","print(\"No. of character comparisons: %d\" % num_char_compar)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"5IzV_9EYjASG","executionInfo":{"status":"ok","timestamp":1656618633180,"user_tz":-60,"elapsed":10,"user":{"displayName":"Sasha malkani","userId":"10596222964619627791"}},"outputId":"f29070da-3c45-47ca-9fa8-68f519462aeb"},"execution_count":57,"outputs":[{"output_type":"stream","name":"stdout","text":["Exact occurrences: 56922\n","No. of alignments: 127974\n","No. of character comparisons: 165191\n"]}]},{"cell_type":"markdown","source":["### Functions:\n","##### 1. naive_2mm (naive exact match with 2 mismatches)\n","##### 2. approximate match\n","##### 3. approximate_match_subseq\n","##### 4. Class Object Index\n","##### 5. Class Object SubseqIndex"],"metadata":{"id":"VmpVq7Bf3Pp1"}},{"cell_type":"code","source":["import bisect\n","\n","\n","def naive_2mm(p, t):\n","    occurrences = []\n","    for i in range(len(t) - len(p) + 1):  # loop over alignments\n","        count_mismatch = 0\n","        for j in range(len(p)):  # loop over characters\n","            if t[i+j] != p[j]:  # compare characters\n","                count_mismatch += 1\n","        if count_mismatch <= 2:\n","            occurrences.append(i)  # all chars matched; record\n","    return occurrences\n","\n","\n","def approximate_match(p, t, n):\n","    segment_length = int(round(len(p) / (n + 1)))\n","    all_matches = set()\n","    p_idx = Index(t, segment_length)\n","    idx_hits = 0\n","    for i in range(n + 1):\n","        start = i * segment_length\n","        end = min((i + 1) * segment_length, len(p))\n","        matches = p_idx.query(p[start:end])\n","\n","        # Extend matching segments to see if whole p matches\n","        for m in matches:\n","            idx_hits += 1\n","            if m < start or m - start + len(p) > len(t):\n","                continue\n","\n","            mismatches = 0\n","\n","            for j in range(0, start):\n","                if not p[j] == t[m - start + j]:\n","                    mismatches += 1\n","                    if mismatches > n:\n","                        break\n","            for j in range(end, len(p)):\n","                if not p[j] == t[m - start + j]:\n","                    mismatches += 1\n","                    if mismatches > n:\n","                        break\n","\n","            if mismatches <= n:\n","                all_matches.add(m - start)\n","    return list(all_matches), idx_hits\n","\n","\n","def approximate_match_subseq(p, t, n, ival):\n","    segment_length = int(round(len(p) / (n + 1)))\n","    all_matches = set()\n","    p_idx = SubseqIndex(t, segment_length, ival)\n","    idx_hits = 0\n","    for i in range(n + 1):\n","        start = i\n","        matches = p_idx.query(p[start:])\n","\n","        # Extend matching segments to see if whole p matches\n","        for m in matches:\n","            idx_hits += 1\n","            if m < start or m - start + len(p) > len(t):\n","                continue\n","\n","            mismatches = 0\n","\n","            for j in range(0, len(p)):\n","                if not p[j] == t[m - start + j]:\n","                    mismatches += 1\n","                    if mismatches > n:\n","                        break\n","\n","            if mismatches <= n:\n","                all_matches.add(m - start)\n","    return list(all_matches), idx_hits\n","\n","\n","# class Index is provided by instructor Ben Langmead\n","class Index(object):\n","    \"\"\" Holds a substring index for a text T \"\"\"\n","\n","    def __init__(self, t, k):\n","        \"\"\" Create index from all substrings of t of length k \"\"\"\n","        self.k = k  # k-mer length (k)\n","        self.index = []\n","        for i in range(len(t) - k + 1):  # for each k-mer\n","            self.index.append((t[i:i+k], i))  # add (k-mer, offset) pair\n","        self.index.sort()  # alphabetize by k-mer\n","\n","    def query(self, p):\n","        \"\"\" Return index hits for first k-mer of p \"\"\"\n","        kmer = p[:self.k]  # query with first k-mer\n","        i = bisect.bisect_left(self.index, (kmer, -1))  # binary search\n","        hits = []\n","        while i < len(self.index):  # collect matching index entries\n","            if self.index[i][0] != kmer:\n","                break\n","            hits.append(self.index[i][1])\n","            i += 1\n","        return hits\n","\n","\n","class SubseqIndex(object):\n","    \"\"\" Holds a subsequence index for a text T \"\"\"\n","\n","    def __init__(self, t, k, ival):\n","        \"\"\" Create index from all subsequences consisting of k characters\n","            spaced ival positions apart.  E.g., SubseqIndex(\"ATAT\", 2, 2)\n","            extracts (\"AA\", 0) and (\"TT\", 1). \"\"\"\n","        self.k = k  # num characters per subsequence extracted\n","        self.ival = ival  # space between them; 1=adjacent, 2=every other, etc\n","        self.index = []\n","        self.span = 1 + ival * (k - 1)\n","        for i in range(len(t) - self.span + 1):  # for each subseq\n","            self.index.append((t[i:i + self.span:ival], i))  # +(subseq, offset)\n","        self.index.sort()  # alphabetize by subseq\n","\n","    def query(self, p):\n","        \"\"\" Return index hits for first subseq of p \"\"\"\n","        subseq = p[:self.span:self.ival]  # query with first subseq\n","        i = bisect.bisect_left(self.index, (subseq, -1))  # binary search\n","        hits = []\n","        while i < len(self.index):  # collect matching index entries\n","            if self.index[i][0] != subseq:\n","                break\n","            hits.append(self.index[i][1])\n","            i += 1\n","        return hits"],"metadata":{"id":"i3J8-bqg43kv","executionInfo":{"status":"ok","timestamp":1656618633181,"user_tz":-60,"elapsed":7,"user":{"displayName":"Sasha malkani","userId":"10596222964619627791"}}},"execution_count":58,"outputs":[]},{"cell_type":"markdown","source":["### Repeat (ALU) sequences (using Naive Exact Matching)\n","##### How many times does the following sequence appear in Human Chromosome 1?\n","##### (allowing for up to 2 substitutions)\n","##### 'GGCGCGGTGGCTCACGCCTGTAAT'"],"metadata":{"id":"ipCqmdQR5KeM"}},{"cell_type":"code","source":["p = 'GGCGCGGTGGCTCACGCCTGTAAT'\n","\n","# Question 4\n","# Using naive_2mm to check the result.\n","print(\"Occurence with up to 2 substitutions using Naive Exact:\",\n","      len(naive_2mm(p, t)))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"hLWXsWny5-P6","executionInfo":{"status":"ok","timestamp":1656618635749,"user_tz":-60,"elapsed":2191,"user":{"displayName":"Sasha malkani","userId":"10596222964619627791"}},"outputId":"05e95282-c66c-4fcc-ecc0-a41f773e39fb"},"execution_count":59,"outputs":[{"output_type":"stream","name":"stdout","text":["Occurence with up to 2 substitutions using Naive Exact: 19\n"]}]},{"cell_type":"markdown","source":["### Repeat (ALU) sequences (using Boyer-Moore)\n","##### How many times does the following sequence appear in Human Chromosome 1?\n","##### (allowing for up to 2 substitutions)\n","##### 'GGCGCGGTGGCTCACGCCTGTAAT'"],"metadata":{"id":"LUKWhBvZ8CB8"}},{"cell_type":"code","source":["matches, hit_num = approximate_match(p, t, 2)\n","\n","# Question 4\n","print(\"Occurence with up to 2 substitutions using Boyer Moore:\", len(matches))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"hBtDa69d8aQn","executionInfo":{"status":"ok","timestamp":1656618637925,"user_tz":-60,"elapsed":2180,"user":{"displayName":"Sasha malkani","userId":"10596222964619627791"}},"outputId":"05271aab-18d5-4f61-fdb2-692f53539a02"},"execution_count":60,"outputs":[{"output_type":"stream","name":"stdout","text":["Occurence with up to 2 substitutions using Boyer Moore: 19\n"]}]},{"cell_type":"markdown","source":["### Number of Index Hits"],"metadata":{"id":"nE9pxmrO9zgs"}},{"cell_type":"code","source":["# Question 5\n","print(\"Total index hits with up to 2 substitutions:\", hit_num)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Kivo22It96wu","executionInfo":{"status":"ok","timestamp":1656618637926,"user_tz":-60,"elapsed":7,"user":{"displayName":"Sasha malkani","userId":"10596222964619627791"}},"outputId":"9490b9a7-af68-4696-eb5b-fa46fcb87115"},"execution_count":61,"outputs":[{"output_type":"stream","name":"stdout","text":["Total index hits with up to 2 substitutions: 90\n"]}]},{"cell_type":"markdown","source":["### Number of Index Hits using subsequences"],"metadata":{"id":"SixkaSmC--RU"}},{"cell_type":"code","source":["# Question 6\n","print(\"Total index hits with up to 2 substitutions, using subsequences:\", \n","      approximate_match_subseq(p, t, 2, 3)[1])\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"K2Yp7S7M_J-l","executionInfo":{"status":"ok","timestamp":1656619261521,"user_tz":-60,"elapsed":2439,"user":{"displayName":"Sasha malkani","userId":"10596222964619627791"}},"outputId":"2d37cb4f-9afc-4eff-9096-6cf82c4918ac"},"execution_count":67,"outputs":[{"output_type":"stream","name":"stdout","text":["Total index hits with up to 2 substitutions, using subsequences: 79\n"]}]}]}